src/txHashBuilder.c

```diff
@@ -288,20 +292,347 @@ void txHashBuilder_addCertificate_delegation(
+	builder->poolCertificateData.remainingRelays--;
+
+	ASSERT(dnsName != NULL);
+	ASSERT(dnsNameSize > 0);
+	ASSERT(dnsNameSize <= DNS_NAME_MAX_LENGTH);
+
+	// Array(2)[
+	//   Unsigned[2]
+	//   Text[dnsName]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, 2);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, dnsNameSize);
+			BUILDER_APPEND_DATA(dnsName, dnsNameSize);
+		}
+	}
+}
+
+// enter empty owners or relays if none were received
+static void addPoolMetadata_updateState(tx_hash_builder_t* builder)
+{
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_PARAMS:
+		// skipping owners is only possible if none were expected
+		ASSERT(builder->poolCertificateData.remainingOwners == 0);
+		txHashBuilder_addPoolRegistrationCertificate_enterOwners(builder);
+
+	// intentional fallthrough
+
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_OWNERS:
+		// skipping relays is only possible if none were expected
+		ASSERT(builder->poolCertificateData.remainingRelays == 0);
+		txHashBuilder_addPoolRegistrationCertificate_enterRelays(builder);
+
+	// intentional fallthrough
+
+	case TX_HASH_BUILDER_IN_CERTIFICATES_POOL_RELAYS:
+		// all relays should have been received
+		ASSERT(builder->poolCertificateData.remainingRelays == 0);
+		break; // we want to be here
+
+	default:
+		ASSERT(false);
+	}
+
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES_POOL_METADATA;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata(
+        tx_hash_builder_t* builder,
+        const uint8_t* url, size_t urlSize,
+        const uint8_t* metadataHash, size_t metadataHashSize
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata: %d", builder->state);
+
+	// we allow this to be called immediately after pool params have been added
+	// if there are no owners or relays in the tx
+	addPoolMetadata_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_METADATA);
+
+	// Array(2)[
+	//   Tstr[url]
+	//   Bytes[metadataHash]
+	// ]
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_TEXT, urlSize);
+			BUILDER_APPEND_DATA(url, urlSize);
+		}
+		{
+			BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, metadataHashSize);
+			BUILDER_APPEND_DATA(metadataHash, metadataHashSize);
+		}
+	}
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES;
+}
+
+void txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata_null(
+        tx_hash_builder_t* builder
+)
+{
+	TRACE("txHashBuilder_addPoolRegistrationCertificate_addPoolMetadata_null: %d", builder->state);
+
+	addPoolMetadata_updateState(builder);
+	ASSERT(builder->state == TX_HASH_BUILDER_IN_CERTIFICATES_POOL_METADATA);
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_NULL, 0);
+	}
+	builder->state = TX_HASH_BUILDER_IN_CERTIFICATES;
+}
+
+static void txHashBuilder_assertCanLeaveCertificates(tx_hash_builder_t* builder)
+{
 	if (builder->state == TX_HASH_BUILDER_IN_CERTIFICATES) {
```

Comment:
I think I like the `switch` variation you introduced in addPoolMetadata_updateState nicer to read